src/libhirte/test/common/network/get_address_test.c

```diff
@@ -7,88 +7,102 @@
 #include "libhirte/common/network.h"
 #include "libhirte/common/string-util.h"
 
-/* test cases with invalid or unresolvable domains */
-bool test_get_address_failure(const char *domain) {
-        _cleanup_free_ char *ip = NULL;
-        int result = get_address(domain, &ip);
-        if (result >= 0) {
-                fprintf(stdout,
-                        "FAILED: get_address('%s') - Expected a failure (result < 0), but got %d\n",
-                        domain,
-                        result);
-                return false;
-        }
-        if (ip != NULL) {
-                fprintf(stdout, "FAILED: get_address('%s') - Expected null, but got non-null (%s)\n", domain, ip);
-                return false;
-        }
-        return true;
-}
+typedef struct get_address_test_params {
+        char *hostname;
+        bool use_ipv4;
+        char *expected_resolved_addresses;
+        int expected_result;
+} get_address_test_params;
 
-/* test cases where domain already contains an IPv4 or IPv6 */
-bool test_get_address_success(const char *domain) {
-        _cleanup_free_ char *ip = NULL;
-        int result = get_address(domain, &ip);
-        if (result != 0) {
-                fprintf(stdout,
-                        "FAILED: get_address('%s') - Expected a success (result == 0), but got %d (%s)\n",
-                        domain,
-                        result,
-                        strerror(-result));
-                return false;
+
+long unsigned int test_data_index = 0;
+struct get_address_test_params test_data[] = {
+        { NULL, true, NULL, -EINVAL },
+        { ".", true, NULL, -1 },
+        { "redhat", true, NULL, -1 },
+        { "?10.10.10.3", true, NULL, -1 },
+        { "192.168.1.", true, NULL, -1 },
+        { "8.8.8.8", true, NULL, 0 },
+        { "localhost.localdomain", true, "127.0.0.1", 0 },
+};
+
+int getaddrinfo_mock(
+                UNUSED const char *name,
+                UNUSED const char *service,
+                UNUSED const struct addrinfo *req,
+                struct addrinfo **pai) {
+
+        void *offset;
+
+        get_address_test_params current_test_case = test_data[test_data_index];
+        if (current_test_case.expected_result < 0) {
+                *pai = NULL;
+                return current_test_case.expected_result;
         }
-        if (ip != NULL) {
-                fprintf(stdout, "FAILED: get_address('%s') - Expected null, but got non-null (%s)\n", domain, ip);
-                return false;
+
+        /*
+         * To emulate what getaddrinfo does, allocate both sockaddr_in and
+         * addrinfo together so freeaddrinfo() will free the entire thing.
+         */
+        *pai = malloc0(sizeof(struct addrinfo) + sizeof(struct sockaddr_in6));
```

Comment:
This doesn't make sense. `freeaddrinfo` will probably call `free` on the `pai` pointer, so all memory is freed - including the subsequently set `sockaddr_in`. The subsequent approach with the `offset` works because of that - using the allocated memory here - and simply calling `free` on the `pai` pointer in the get_address code would probably suffice.